core/base/segmented_range.hpp

```diff
@@ -0,0 +1,372 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_CORE_BASE_SEGMENTED_RANGE_HPP_
+#define GKO_CORE_BASE_SEGMENTED_RANGE_HPP_
+
+
+#include <iterator>
+#include <type_traits>
+
+#include "core/base/index_range.hpp"
+#include "core/base/iterator_factory.hpp"
+
+
+namespace gko {
+
+
+/**
+ * Represents a range of indices that is segmented into contiguous segments.
+ * Each segment has the shape `[begin, end)`, i.e. it is a half-open interval.
+ *
+ * @tparam IndexType  the type of indices used to represent the segments.
+ */
+template <typename IndexType>
+class segmented_range {
+public:
+    using index_type = IndexType;
+    using index_iterator_type = index_iterator<index_type>;
```

Comment:
not really related to this pr, but maybe a better name for `index_iterator` would be `interval_iterator`.

Reply:
I think we should have clarity on what the iterator typename represents - is it an iterator _returning_ an index or an interator _pointing into_ an interval. Both make sense, but I chose the first one initially.

Reply:
Thinking about it, I now would also prefer how it is currently done. So I would prefer an iterator being named after what it returns.